### PR TITLE
[SPARK-39860][SQL] More expressions should extend Predicate

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR makes more expressions extend `Predicate`.

### Why are the changes needed?

To avoid adding a useless cast:
https://github.com/apache/spark/blob/46175d1362062035fb93f87f25d61a9b711359ab/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L648-L660

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BloomFilterMightContain.scala [hunk 1]
```diff
@@ -40,13 +40,12 @@ import org.apache.spark.util.sketch.BloomFilter
  */
 case class BloomFilterMightContain(
     bloomFilterExpression: Expression,
-    valueExpression: Expression) extends BinaryExpression {
+    valueExpression: Expression) extends BinaryExpression with Predicate {
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BloomFilterMightContain.scala:43]: AFAK, although the `dataType` of `BloomFilterMightContain`, but users cannot use it as predicate directly.

**Comment by wangyum** [comment]
Example:
```scala
spark.sql("CREATE TABLE t1(a int, b string) using parquet")
spark.sql("SELECT a, max(b) AS max_b FROM t1 GROUP BY 1 HAVING max_b LIKE '%b%' ").explain(true)
```

**Review by beliefer — COMMENTED** [review]
LGTM expect one comment !

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BloomFilterMightContain.scala (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BloomFilterMightContain.scala:43]: Yes.

**Comment by wangyum** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
can we check all the usages of `Predicate`? I'm afraid Spark may have more assumptions about `Predicate`, not just avoid casting to boolean.

cc @sigmod

**Comment by wangyum** [comment]
All the value read usages:
image

https://github.com/apache/spark/blob/8628c15d176a8084b695bbb6ca841831e6fb75d4/sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala#L161-L162
https://github.com/apache/spark/blob/fc157d8d4908f473ce32c2696bb9013dcaecd705/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L787-L790
https://github.com/apache/spark/blob/fc157d8d4908f473ce32c2696bb9013dcaecd705/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L655-L658

**Review by cloud-fan — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master.

**Pull request closed without merging**